src/color_laws.jl

```diff
@@ -228,3 +230,143 @@ function gcc09_invum(x::Real, Rv::Real)
 
     return a + b / Rv
 end
+
+# Shape models used by F99 and F04
+function _curve_F99_method(
+    x::Real,
+    Rv::Real,
+    c1::Real,
+    c2::Real,
+    c3::Real,
+    c4::Real,
+    x0::Real,
+    gamma::Real,
+    optnir_axav_x::Vector{<:Real},
+    optnir_axav_y::Vector{<:Real},
+    )
+
+    # x value above which FM90 parametrization used
+    x_cutval_uv = 10000.0 / 2700.0
+
+    # required UV points for spline interpolation
+    x_splineval_uv = 10000.0 ./ [2700.0, 2600.0]
+
+    # add in required spline points, otherwise just spline points
+    if x >= x_cutval_uv
+        xuv = vcat(x_splineval_uv, x)
+    else
+        xuv = x_splineval_uv
+    end
+
+    # FM90 model and values
+    fm90_model = FM90(c1=c1, c2=c2, c3=c3, c4=c4, x0=x0, gamma=gamma)
+    # evaluate model and get results in A(x)/A(V)
+    axav_fm90 = fm90_model.(10000.0 ./ xuv) / Rv .+ 1.0 # Expects xuv in Å
+
+    # ignore the spline points
+    if x >= x_cutval_uv
+        axav = last(axav_fm90)
+    end
+
+    # **Optical Portion**
+    #   using cubic spline anchored in UV, optical, and IR
+
+    # optical/NIR points in input x
+    if x < x_cutval_uv
+        # save spline points
+        y_splineval_uv = axav_fm90
+
+        # spline points
+        x_splineval_optir = optnir_axav_x
+
+        # determine optical/IR values at spline points
+        y_splineval_optir = optnir_axav_y
+
+        spline_x = vcat(x_splineval_optir, x_splineval_uv)
+        spline_y = vcat(y_splineval_optir, y_splineval_uv)
+        spl = Spline1D(spline_x, spline_y, k=3)
+        axav = spl(x)
+    end
+
+    # return A(x)/A(V)
+    return axav
+end
+
+"""
+    F99(;Rv=3.1)
+
+Fitzpatrick (1999) dust law.
+
+# References
+[Fitzpatrick (1999)](https://ui.adsabs.harvard.edu/abs/1999PASP..111...63F/)
+"""
+@with_kw struct F99 <: ExtinctionLaw
+    Rv::Float64 = 3.1
+end
+
+function (law::F99)(wave::T) where T
+    checkbounds(law, wave) || return zero(float(T))
+    x = aa_to_invum(wave)
+    return f99_invum(x, law.Rv)
+end
+
+bounds(::Type{F99}) = (1000.0, 33333.3)
+
+"""
+    DustExtinction.f99_invum(x, Rv)
+
+The algorithm used for the [`F99`](@ref) extinction law, given inverse microns and Rv. For more information, seek the original paper.
+"""
+function f99_invum(x::Real, Rv::Real)
+    if !(0.3 <= x <= 10.0)
+        error("out of bounds of F99, support is over $(bounds(F99)) angstrom")
+    end
```

Comment:
I think this is redundant because we already have checked bounds in line number-308.

Reply:
I was wondering about that too. Removing this check seems to cause the following [tests](https://github.com/icweaver/DustExtinction.jl/blob/F99/test/color_laws.jl#L271-L274) to fail (copied below):

```
bad_xs = [0.2, 11.0]
@test @inferred(map(law, bad_xs)) == zeros(length(bad_xs))
@test_throws ErrorException f99_invum(bad_xs[1], rv)
@test_throws ErrorException f99_invum(bad_xs[2], rv)
```

I think the bounds check on line number-308 is only for the top level interface (e.g. `F99(<Rv value>)(<wavelength in Å>`)), while this test is for the lower level input (e.g. `f99_invum(<wavenumber in 1/microns>, <Rv value>`))? 

Definitely agree though that this seems a little redundant.

Reply:
Do astronomers use `invum` function directly? This is not exposed directly to the user in our package. Thoughts? @mileslucas @giordano

Reply:
Yes, this is redundant, good catch!

Reply:
I see the code as two parts: the public facing API meant for consumption, as well as the internal implementations which are meant to more rigorously represent the original mathematical derivations. This is why the public API returns 0 extinction instead of erroring outside the limits: because nothing makes ~me~astronomers grouchier than having to precisely mask your spectra to match rough empirical limits. However, the `_invum` functions pretty directly translate to the mathematical definitions and I want them to be explicit in their behaviors around the bounds.

As for the bound discrepancies- I would just choose one and keep it consistent between this and `bounds`.

Reply:
Ok, just dropped! d81b97fd8b72bb570ee44f91be64bbfe4e6ee091

I guess these should also be removed from the other models in another PR?

Reply:
Added a [post](https://github.com/JuliaAstro/DustExtinction.jl/issues/10#issuecomment-655706169) in the original issue to help keep track of things

Reply:
Just closing the loop on some of our threads. The test as been put back re: https://github.com/JuliaAstro/DustExtinction.jl/pull/30#discussion_r451765431